src/target/riscv/riscv-013.c

```diff
@@ -1521,300 +1217,38 @@ static int examine(struct target *target)
-	if (get_field(dmstatus, DMI_DMSTATUS_ANYRUNNING)) {
-		should_attempt_resume = true;
-		LOG_DEBUG("Hart currently running. Requesting halt.\n");
-		dmi_write(target, DMI_DMCONTROL, DMI_DMCONTROL_HALTREQ |
-				DMI_DMCONTROL_DMACTIVE);
-		for (unsigned i = 0; i < 256; i++) {
-			dmcontrol = dmi_read(target, DMI_DMCONTROL);
-			dmstatus = dmi_read(target, DMI_DMSTATUS);
-			if (get_field(dmstatus, DMI_DMSTATUS_ALLHALTED))
-				break;
-		}
-		if (!get_field(dmstatus, DMI_DMSTATUS_ALLHALTED)) {
-			LOG_ERROR("hart didn't halt; dmstatus=0x%x", dmstatus);
-			return ERROR_FAIL;
-		}
-	}
+	/* Halt every hart so we can probe them. */
+	riscv_halt_all_harts(target);
 
-	// TODO: do this using Quick Access, if supported.
+	/* Examines every hart, first checking XLEN. */
+	for (int i = 0; i < riscv_count_harts(target); ++i) {
+		RISCV_INFO(r);
+		riscv_set_current_hartid(target, i);
 
-	riscv_info_t *generic_info = (riscv_info_t *) target->arch_info;
-	if (abstract_read_register(target, NULL, S0, 128) == ERROR_OK) {
-		generic_info->xlen = 128;
-	} else if (abstract_read_register(target, NULL, S0, 64) == ERROR_OK) {
-		generic_info->xlen = 64;
-	} else if (abstract_read_register(target, NULL, S0, 32) == ERROR_OK) {
-		generic_info->xlen = 32;
-	} else {
-		LOG_ERROR("Failed to discover size using abstract register reads.");
-		return ERROR_FAIL;
-	}
-
-	LOG_DEBUG("Discovered XLEN is %d", xlen(target));
-
-	// Update register list to match discovered XLEN.
-	update_reg_list(target);
-
-	if (register_read_direct(target, &info->misa, REG_MISA) != ERROR_OK) {
-		LOG_ERROR("Failed to read misa.");
-		return ERROR_FAIL;
-	}
-
-	if (should_attempt_resume) {
-		LOG_DEBUG("Resuming hart.\n");
-		// Resume if the hart had been running.
-		dmi_write(target, DMI_DMCONTROL, DMI_DMCONTROL_DMACTIVE |
-				DMI_DMCONTROL_RESUMEREQ);
-		for (unsigned i = 0; i < 256; i++) {
-			dmcontrol = dmi_read(target, DMI_DMCONTROL);
-			dmstatus = dmi_read(target, DMI_DMSTATUS);
-			if (get_field(dmstatus, DMI_DMSTATUS_ALLRUNNING))
-				break;
-		}
-		if (!get_field(dmstatus, DMI_DMSTATUS_ALLRUNNING)) {
-			LOG_ERROR("hart didn't resume; dmstatus=0x%x", dmstatus);
+		if (abstract_read_register(target, NULL, S0, 128) == ERROR_OK) {
+			r->xlen[i] = 128;
+		} else if (abstract_read_register(target, NULL, S0, 64) == ERROR_OK) {
+			r->xlen[i] = 64;
+		} else if (abstract_read_register(target, NULL, S0, 32) == ERROR_OK) {
+			r->xlen[i] = 32;
+		} else {
+			LOG_ERROR("Failed to discover size using abstract register reads.");
 			return ERROR_FAIL;
 		}
 	}
 
-	info->never_halted = true;
-
-	int result = riscv013_poll(target);
-	if (result != ERROR_OK) {
-		return result;
-	}
+	/* FIXME: Are there 2 triggers? */
+	info->trigger_count = 2;
 
+	/* Resumes all the harts, so the debugger can later pause them. */
+	riscv_resume_all_harts(target);
 	target_set_examined(target);
-	LOG_INFO("Examined RISCV core; XLEN=%d, misa=0x%" PRIx64, xlen(target), info->misa);
```

Comment:
I'm going to miss this printf. Actually don't some of the tests look for this string to pass?